Rework Quickstart guide (#1193)
scality/metalk8s: docs/quickstart/introduction.rst

Introduction
============

> Deploying a MetalK8s cluster, while simple, will be facilitated by

Reviewer: Not sure this paragraph is necessary.

an introductory explanation of a few concepts and an outline of the deployment
procedure that will ensue.

This guide will cover many aspects while not diving too much into the details.

> Links to relevant sections of the complete documentation are included when

Reviewer: Ditto for this sentence, the glossary link is already in the sidebar.

useful, and a :doc:`/glossary` is provided.

Concepts
^^^^^^^^
Although being familiar with
`Kubernetes concepts <https://kubernetes.io/docs/concepts/>`_
is recommended, the necessary concepts to grasp before installing a MetalK8s
cluster are presented here.

Control-plane and workload-plane
""""""""""""""""""""""""""""""""

.. _`API Server`:
      https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/
.. _Scheduler:
      https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/
.. _`Controller Manager`:
      https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/
.. _etcd: https://etcd.io/

> 

Reviewer: In keeping with the idea of having a workable offline doc, these should be glossary terms. Minimal description + link to outside doc go there.

This dichotomy is central to MetalK8s, and often referred to in other
Kubernetes concepts.

> The **control-plane** is the set of machines (called :term:`nodes <Node>`) and

Reviewer: I would introduce the `node` term in the previous paragraph, since control-plane/workload-plane have the most sense on nodes.

the services running there that make up the essential Kubernetes functionality
for running containerized applications, managing declarative objects, and
providing authentication/authorization to end-users as well as services.
The main components making up a Kubernetes control-plane are:

- `API Server`_

> - Scheduler_

Reviewer 2: Why `Scheduler` is not surrounded by backquote?

Author: They are not needed, only when the link contains spaces or special characters. I can add them for consistency if you'd like.

Reviewer 2: I would do it for consistency yeah.

- `Controller Manager`_

> The **workload-plane** indicates the ensemble of nodes where applications

Reviewer: Switch `ensemble` to `group`, `set`... Ensemble mostly makes sense musically :notes:

will be deployed via Kubernetes objects, using services provided by the
**control-plane**.

> Note that nodes can belong to both planes, so that one can run applications

Reviewer: Should be a `.. note::`, and `s/can belong/may belong`

alongside the control-plane services.

Control-plane nodes often are responsible for providing storage for
`API Server`_, by running `etcd`_. This responsibility may be offloaded to

> other nodes from the workload-plane.

Reviewer: ```suggestion
nodes not belonging to the control-plane.
```

Author: No, I want to say that it is acceptable to run `etcd` on workload-plane nodes (i.e. not setting the `etcd` taint).

Reviewer: Let's be explicit on that point, then. Something like

> This responsibility may be offloaded to workload-plane nodes, without the `etcd` taint.

Node roles
""""""""""

Determining a :term:`Node` responsibilities is achieved using **roles**.
Roles are stored in :term:`Node manifests <Node manifest>` using labels, of the
form ``node-role.kubernetes.io/<role-name>: ''``.

> MetalK8s uses five different **roles**, that may be combined freely:

Reviewer: ```suggestion
MetalK8s uses five different **roles** that may be combined freely:
```

> ``node-role.kubernetes.io/master``

Reviewer: Let's use `marks` instead of `defines` etc... on all these definitions for consistency

Author: > I meant to have a comment with that review ><
> 
> As far as the structure goes it's really well laid out in this PR, nice work!

Thanks!

> Most of my comments are wording nitpicks etc..., feel free to `wontdo` any of them.

Actually, will do almost all of them, thanks for the help on removing `you` ocurrences :)

  The ``master`` role defines control-plane membership. Control-plane services
  (see above) can only be scheduled on ``master`` nodes.

``node-role.kubernetes.io/etcd``
  The ``etcd`` role marks a node running `etcd`_ for storage of API Server.

``node-role.kubernetes.io/node``
  This role defines a workload-plane node. It is included implicitly by all
  other roles.

``node-role.kubernetes.io/infra``
  The ``infra`` role is specific to MetalK8s. It serves for marking nodes where
  non-critical services provided by the cluster (monitoring stack, UIs, etc.)
  are running.

``node-role.kubernetes.io/bootstrap``
  This marks the :term:`Bootstrap node`. This node is unique in the cluster,

>   and runs a few specific services:

Reviewer: ```suggestion
  and is solely responsible for the following services:
```

>   - A package repository for Nodes to install from

Reviewer: ```suggestion
  - An RPM package repository used by Nodes joining the cluster
```

Author: Since the repository can be used after joining (upgrade/downgrade), I'll go for:

```suggestion
  - An RPM package repository used by Nodes
```

Reviewer: LGTM like this - minor suggestion: maybe `cluster members` instead of `Nodes`?

  - An OCI registry for :term:`Pods <Pod>` images

>   - A Salt Master, deployed with SaltAPI

Reviewer: ```suggestion
  - A Salt Master service and its associated SaltAPI service
```

Also, Salt Master and SaltAPI could be `:term:`s

Author: Why bother adding `service`? Agreed for the glossary definitions

Reviewer: Yeah let's not have `service`, but keep the association.

>   In practice, it will be used in conjunction with the ``master`` and ``etcd``

Reviewer: ```suggestion
  In practice, this role will be used in conjunction with the ``master`` and ``etcd``
```

  roles.

.. _quickstart-intro-networks:

Networks
""""""""

.. todo::

   - need physical networks for control-plane and workload-plane (may be the
     same)
   - define how each node in the cluster needs an IP for each network
   - mention virtual networks for Pods and Services, managed by the CNI

Installation plan
^^^^^^^^^^^^^^^^^

In this guide, the depicted installation procedure is for a medium sized

> cluster, using three control-plane nodes and two worker nodes. Hints for

Reviewer: No need for the "Hints [...]" sentence IMO.

deploying a smaller architecture will be included, though sparse. Refer to
the :doc:`/installation-guide/index` for extensive explanations of possible
cluster architectures.

> Here is how it will happen:

Reviewer: ```suggestion
The installation process can be broken down in 3 separate steps:
```
